src/luaotfload-harf-define.lua

```diff
@@ -0,0 +1,372 @@
+          italic = -extents.x_bearing
+        end
+      end
+      glyphs[gid] = {
+        width  = width,
+        height = height or ascender,
+        depth  = -(depth or descender),
+        italic = italic or 0,
+      }
+    end
+
+    local unicodes = hbface:collect_unicodes()
+    local characters = {}
+    local nominals = {}
+    for _, uni in next, unicodes do
+      local glyph = hbfont:get_nominal_glyph(uni)
+      characters[uni] = glyph
+      nominals[glyph] = uni
+    end
+
+    local xheight, capheight = 0, 0
+    if hasos2 then
+      local os2 = hbface:get_table(os2tag)
+      local length = os2:get_length()
+      local data = os2:get_data()
+      if length >= 96 and string.unpack(">H", data) > 1 then
+        -- We don’t need much of the table, so we read from hard-coded offsets.
+        xheight = string.unpack(">H", data, 87)
+        capheight = string.unpack(">H", data, 89)
+      end
+    end
+
+    if xheight == 0 then
+      local gid = characters[120] -- x
+      if gid then
+        xheight = glyphs[gid].height
+      else
+        xheight = ascender / 2
+      end
+    end
+
+    if capheight == 0 then
+      local gid = characters[88] -- X
+      if gid then
+        capheight = glyphs[gid].height
+      else
+        capheight = ascender
+      end
+    end
+
+    data = {
+      gid_offset = 0x120000,
+      face = hbface,
+      font = hbfont,
+      upem = upem,
+      fonttype = fonttype,
+      space = space,
+      xheight = xheight,
+      capheight = capheight,
+      slant = slant,
+      glyphs = glyphs,
+      nominals = nominals,
+      unicodes = characters,
+      psname = hbface:get_name(hb.ot.NAME_ID_POSTSCRIPT_NAME),
+      fullname = hbface:get_name(hb.ot.NAME_ID_FULL_NAME),
+      haspng = hbface:ot_color_has_png(),
+      loaded = {}, -- Cached loaded glyph data.
+    }
+
+    hbfonts[key] = data
+    return data
+  end
+end
+
+-- Drop illegal characters from PS Name, per the spec
+-- https://docs.microsoft.com/en-us/typography/opentype/spec/name#nid6
+local function sanitize(psname)
+  return psname:gsub('[][\0-\32\127-\255(){}<>/%%]', '-')
+end
+
+-- Ligatures. The value is a character "ligature" table as described in the
+-- manual.
+local tlig ={
+  [0x2013] = { [0x002D] = { char = 0x2014 } }, -- [---]
+  [0x002D] = { [0x002D] = { char = 0x2013 } }, -- [--]
+  [0x0060] = { [0x0060] = { char = 0x201C } }, -- [``]
+  [0x0027] = { [0x0027] = { char = 0x201D } }, -- ['']
+  [0x0021] = { [0x0060] = { char = 0x00A1 } }, -- [!`]
+  [0x003F] = { [0x0060] = { char = 0x00BF } }, -- [?`]
+  [0x002C] = { [0x002C] = { char = 0x201E } }, -- [,,]
+  [0x003C] = { [0x003C] = { char = 0x00AB } }, -- [<<]
+  [0x003E] = { [0x003E] = { char = 0x00BB } }, -- [>>]
+}
+
+local function scalefont(data, spec)
+  local size = spec.size
+  local features = spec.features.normal
+  features.mode = 'plug'
+  features.features = 'harf'
+  fonts.constructors.checkedfeatures("otf", features)
```

Comment:
Is this the code that adds script-specific features? If so it shouldn’t be done here as HarfBuzz maintains this internally and manually enabling features that require text analysis by the engine (e.g. Arabic `init`) will activate it globally for the buffer and give the wrong result.

Reply:
This adds default features, not script specific default features. This is necessary to properly interact with some user-defined features. Anyway, this does not interact with HarfBuzz features at all: They are extracted earlier and only look at `features.raw`. That table only contains the features passed by the user, not automatic features.